### PR TITLE
Autograd for `grouped_matmul`

### PR DESCRIPTION
**Pull request by puririshi98**
also includes backward for other in segment_matmul

### CODE REVIEW

File: pyg_lib/csrc/ops/matmul.cpp [hunk 1]
```diff
@@ -52,14 +127,14 @@ class SegmentMatmul : public torch::autograd::Function<SegmentMatmul> {
...
-      // TODO (matthias) get rid of unnecessary `contiguous` here.
-      auto other_t = other.transpose(-2, -1).contiguous();
+      auto other_t = other.transpose(-2, -1);
       input_grad = _segment_matmul(grad_out, ptr, other_t);
     }
     if (torch::autograd::any_variable_requires_grad({other})) {
-      // TODO (matthias) implement backward pass for `other`.
+      variable_list grad_out_list = break_w_ptr(grad_out, ptr);
+      variable_list other_list = break_w_ptr(other, ptr);
+      other_grad = reflatten(_grouped_matmul(grad_out_list, other_list));
```
Comment (rusty1s) [inline pyg_lib/csrc/ops/matmul.cpp:137]: Just use `at::stack(dim=0)` for this.

File: pyg_lib/csrc/ops/matmul.cpp [hunk 2]
```diff
@@ -57,9 +132,10 @@ class SegmentMatmul : public torch::autograd::Function<SegmentMatmul> {
       input_grad = _segment_matmul(grad_out, ptr, other_t);
     }
     if (torch::autograd::any_variable_requires_grad({other})) {
-      // TODO (matthias) implement backward pass for `other`.
+      variable_list grad_out_list = break_w_ptr(grad_out, ptr);
+      variable_list other_list = break_w_ptr(other, ptr);
```
Comment (rusty1s) [inline pyg_lib/csrc/ops/matmul.cpp:136]: `other` can be just split in the first dimension. It is a 3d-tensor.

File: pyg_lib/csrc/ops/matmul.cpp [hunk 3]
```diff
@@ -57,9 +132,10 @@ class SegmentMatmul : public torch::autograd::Function<SegmentMatmul> {
       input_grad = _segment_matmul(grad_out, ptr, other_t);
     }
     if (torch::autograd::any_variable_requires_grad({other})) {
-      // TODO (matthias) implement backward pass for `other`.
+      variable_list grad_out_list = break_w_ptr(grad_out, ptr);
```
Comment (rusty1s) [inline pyg_lib/csrc/ops/matmul.cpp:135]: We can use `split` for this - similar to https://github.com/pyg-team/pyg-lib/blob/master/pyg_lib/csrc/ops/cuda/matmul_kernel.cu#L132-L136.

File: pyg_lib/csrc/ops/matmul.cpp [hunk 4]
```diff
@@ -29,9 +86,27 @@ at::Tensor _segment_matmul(const at::Tensor& input,
...
 }
 
-using torch::autograd::AutogradContext;
-using torch::autograd::Variable;
-using torch::autograd::variable_list;
+std::vector<at::Tensor> break_w_ptr(const at::Tensor& tens,
+                                    const at::Tensor& ptr) {
+  std::vector<at::Tensor> return_list;
+  for (int64_t i = 0; i < ptr.numel(); ++i)
+    return_list.push_back(tens.slice(0, ptr.index({i - 1}).item<int>(),
```
Comment (rusty1s) [inline pyg_lib/csrc/ops/matmul.cpp:93]: This will break since `ptr` is a CUDA tensor.

File: pyg_lib/csrc/ops/matmul.cpp [hunk 5]
```diff
@@ -19,6 +22,60 @@ std::vector<at::Tensor> _grouped_matmul(const std::vector<at::Tensor>& input,
...
+    variable_list input_and_other = concat(input, other);
+    ctx->save_for_backward(input_and_other);
+    ctx->saved_data["input_len"] = (int)input.size();
+    return out;
+  }
+
+  static variable_list backward(AutogradContext* ctx, variable_list grad_outs) {
+    auto input_and_other = ctx->get_saved_variables();
+    int input_len = ctx->saved_data["input_len"].toInt();
+    auto input = get_input(input_and_other, input_len);
```
Comment (rusty1s) [inline pyg_lib/csrc/ops/matmul.cpp:59]: Let's just do this inline? I don't think we need to specify `get_input` and `get_other` here.

File: pyg_lib/csrc/ops/matmul.cpp [hunk 6]
```diff
@@ -57,9 +132,10 @@ class SegmentMatmul : public torch::autograd::Function<SegmentMatmul> {
       input_grad = _segment_matmul(grad_out, ptr, other_t);
     }
     if (torch::autograd::any_variable_requires_grad({other})) {
-      // TODO (matthias) implement backward pass for `other`.
+      variable_list grad_out_list = break_w_ptr(grad_out, ptr);
+      variable_list other_list = break_w_ptr(other, ptr);
+      other_grad = reflatten(_grouped_matmul(grad_out_list, other_list));
```
Comment (rusty1s) [inline pyg_lib/csrc/ops/matmul.cpp:137]: This will still fail since it will not pick up the right dispatcher here, correct?

**Review by rusty1s — CHANGES_REQUESTED** [review]

**Comment by puririshi98** [comment]
tested that forward and backward work for both segment and grouped matmul in pytorch

File: pyg_lib/csrc/ops/matmul.cpp [hunk 7]
```diff
@@ -29,11 +30,55 @@ at::Tensor _segment_matmul(const at::Tensor& input,
...
+
+  static variable_list backward(AutogradContext* ctx, variable_list grad_outs) {
+    auto input_and_other = ctx->get_saved_variables();
+    int input_len = ctx->saved_data["input_len"].toInt();
+    std::vector<at::Tensor> input(input_and_other.begin(),
+                                  input_and_other.begin() + input_len);
+    std::vector<at::Tensor> other(input_and_other.begin() + input_len,
+                                  input_and_other.end());
+    for (size_t i = 0; i < input.size(); ++i)
+      other[i] = other[i].transpose(-2, -1).contiguous();
```
Comment (rusty1s) [inline pyg_lib/csrc/ops/matmul.cpp:65]:
```suggestion
      other[i] = other[i].transpose(-2, -1);
```

File: pyg_lib/csrc/ops/matmul.cpp [hunk 8]
```diff
@@ -29,11 +30,55 @@ at::Tensor _segment_matmul(const at::Tensor& input,
...
+    auto other_grad = _grouped_matmul(grad_outs, other);
+    variable_list input_grad;
+    // For Simplicity:
+    // We assume entire input variable list either requires grad or does not
+    if (torch::autograd::any_variable_requires_grad(input)) {
```
Comment (rusty1s) [inline pyg_lib/csrc/ops/matmul.cpp:70]: Do the same for other?

**Review by rusty1s — APPROVED** [review]
LGTM :)

**Comment by puririshi98** [comment]
```
>>> import torch
>>> import pyg_lib
>>> inputs = torch.randn(8, 16).cuda()
>>> ptr = torch.tensor([0, 5, 8]).cuda()
>>> other = torch.randn((2, 16, 32), requires_grad=True, device='cuda:0')
>>> out = pyg_lib.ops.segment_matmul(inputs, ptr, other)
>>> out.sum().backward()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/conda/lib/python3.8/site-packages/torch/_tensor.py", line 399, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/opt/conda/lib/python3.8/site-packages/torch/autograd/__init__.py", line 173, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
NotImplementedError: There were no tensor arguments to this function (e.g., you passed an empty list of Tensors), but no fallback function is registered for schema pyg::grouped_matmul_kern.  This usually means that this function requires a non-empty list of Tensors, or that you (the operator writer) forgot to register a fallback function. 
```

the dispatching is working properly for segment_matmul now but the grouped_matmul kernel is not dispatching properly. i confirmed with debug prints that the inputs to grouped_matmul_kernel are in fact not empty, so the error message is misleading.

**Comment by puririshi98** [comment]
i found the issue [here](https://gitlab-master.nvidia.com/dl/pytorch/pytorch/-/blob/22.08_jetson/aten/src/ATen/core/op_registration/README.md#L256): 

> * The operator must return a single `Tensor` or multiple tensors as in `(Tensor, Tensor, Tensor)`. It cannot return a list `Tensor[]`, optional `Tensor?` or any primitive types.

**Comment by puririshi98** [comment]
ready to go, even added a little test https://github.com/puririshi98/pyg-lib/blob/back_for_grouped/test/ops/test_cutlass_autograd_ops.py

**Comment by rusty1s** [comment]
I have merged this now. We can continue working on it in later PRs.

**Pull request merged**